dump/mapping/prepbufr_adpupa.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python3
+import os
+import numpy as np
+import time
+import calendar
+from datetime import datetime
+
+import bufr
+from bufr.obs_builder import ObsBuilder, add_main_functions, map_path
+from prepbufr_obs_builder import PrepbufrObsBuilder
+
+MAPPING_PATH = map_path('prepbufr_adpupa.yaml')
+
+
+class AdpupaPrepbufrObsBuilder(PrepbufrObsBuilder):
+    """
+    A builder class to generate ADPUPA observations from ADPUPA prepBUFR input.
+    """
+
+    def __init__(self):
+        super().__init__(MAPPING_PATH, log_name=os.path.basename(__file__))
+
+    def make_obs(self, comm, input_path):
+
+        # Get container from mapping file first
+        self.log.info(f'Get container from bufr')
+        container = super().make_obs(comm, input_path)
+
+        self.log.debug(f'container list (original): {container.list()}')
+
+        self.log.debug(f'Perform DateTime calculation')
+        hrdr = container.get('obsTimeMinusCycleTime')
+        cycle_time = "2021080100"
+        cycleTimeSinceEpoch = np.int64(calendar.timegm(time.strptime(str(int(cycle_time)), '%Y%m%d%H')))
+        dateTime = self._compute_datetime(cycleTimeSinceEpoch, hrdr)
+
+        self.log.debug(f'Perform stationPressure calculation')
+        cat = container.get('prepbufrDataLevelCategory')
+        pob = container.get('pressure')
+        ps = np.full(pob.shape[0], pob.fill_value)
+        ps = np.where(cat == 0, pob, ps)
+
+        self.log.debug(f'Perform stationPressureQM calculation')
+        pqm = container.get('pressureQualityMarker')
+        psqm = np.full(pqm.shape[0], pqm.fill_value)
+        psqm = np.where(cat == 0, pqm, psqm)
+
+        self.log.debug(f'Perform stationPressureError calculation')
+        poe = container.get('pressureError')
+        psoe = np.full(poe.shape[0], poe.fill_value)
+        psoe = np.where(cat == 0, poe, psoe)
+
+        self.log.debug(f'Perform airTemperature and virtualTemperature calculations')
+        tpc = container.get('temperatureEventProgramCode')
+        tob = container.get('airTemperature')
+        tsen = np.full(tob.shape[0], tob.fill_value)
+        tsen = np.where(((tpc >= 1) & (tpc < 8)), tob, tsen)
+        tvo = np.full(tob.shape[0], tob.fill_value)
+        tvo = np.where((tpc == 8), tob, tvo)
```

Comment:
could handle in simular way

Reply:
Could make a function to do this maybe if it cleans things up.

Reply:
@rmclaren done, thanks! Spelled the variables out and added a new function for this to aviod repetition of the codes.